csharp/src/Microsoft.ML.OnnxRuntime/Training/TrainingSession.shared.cs

```diff
@@ -319,14 +338,125 @@ public void OptimizerStep(RunOptions options)
+        /// </summary>
+        /// <param name="inference_model_path">Path where the inference model should be serialized to.</param>
+        /// <param name="graphOutputNames">Names of the outputs that are needed in the inference model.</param>
+        public void ExportModelForInferencing(string inferenceModelPath, IReadOnlyCollection<string> graphOutputNames)
+        {
+            using (var cleanupList = new DisposableList<IDisposable>())
+            {
+                var outputNamesArray = ConvertNamesToUtf8(graphOutputNames, cleanupList);
+                NativeApiStatus.VerifySuccess(NativeTrainingMethods.OrtExportModelForInferencing(
+                    _nativeHandle, NativeOnnxValueHelper.GetPlatformSerializedString(inferenceModelPath),
+                    (UIntPtr)graphOutputNames.Count, outputNamesArray));
+            }
+        }
+
+        /// <summary>
+        /// Returns a contiguous buffer that holds a copy of all training state parameters
+        /// </summary>
+        /// <param name="only_trainable">Whether to only copy trainable parameters or to copy all parameters.</param>
+        public FixedBufferOnnxValue ToBuffer(bool only_trainable)
+        {
+            UIntPtr bufferSize = UIntPtr.Zero;
+            NativeApiStatus.VerifySuccess(NativeTrainingMethods.OrtGetParametersSize(_nativeHandle, out bufferSize, only_trainable));
+
+            float[] bufferMemory = new float[bufferSize.ToUInt64()];
+
+            var memInfo = OrtMemoryInfo.DefaultInstance; // CPU
+            var shape = new long[] {(long)bufferSize.ToUInt64()};
+            var buffer = FixedBufferOnnxValue.CreateFromMemory<float>(memInfo, bufferMemory, Tensors.TensorElementType.Float, shape, (long)bufferSize.ToUInt64() * sizeof(float));
+
+            NativeApiStatus.VerifySuccess(NativeTrainingMethods.OrtCopyParametersToBuffer(_nativeHandle, buffer.Value.Handle, only_trainable));
+
+            return buffer;
+        }
+
+        /// <summary>
+        /// Loads the training session model parameters from a contiguous buffer
+        /// </summary>
+        /// <param name="buffer">Contiguous buffer to load the parameters from.</param>
+        public void FromBuffer(FixedBufferOnnxValue buffer)
+        {
+            if (buffer.OnnxValueType != OnnxValueType.ONNX_TYPE_TENSOR)
+            {
+                throw new ArgumentException("Incorrect buffer received. Expected a tensor buffer.");
+            }
+
+            IntPtr typeAndShapeInfo = IntPtr.Zero;
+            NativeApiStatus.VerifySuccess(NativeMethods.OrtGetTensorTypeAndShape(buffer.Value.Handle, out typeAndShapeInfo));
+            UIntPtr numDimensions = UIntPtr.Zero;
+            NativeApiStatus.VerifySuccess(NativeMethods.OrtGetDimensionsCount(typeAndShapeInfo, out numDimensions));
+            if (numDimensions.ToUInt64() != 1)
+            {
+                throw new ArgumentException("Incorrect buffer received. Expected a contiguous tensor buffer.");
```

Comment:
nit: Include the value received so easier to debug. 

In general if we can include that info in an error message and a user has an issue we have that to work with without having to setup a repro of the issue (if that is even possible). So it might not be overly meaningful to the user, but cut-and-pasted into the error report it's useful to us.